### DISCUSSION TITLE
How to create a mixin for a declarative dataclass model, when the mixin adds a column with a default value?

### DISCUSSION THREAD

**Discussion by Jakkara** [sqlalchemy/sqlalchemy#10105] [Usage Questions] [answered]
I'm trying to wrap my head around how the declarative class mapping works. What I'm trying to achieve is add a `deleted_at` field to a class via a mixin so that it:

- defaults to None
- can be declared in the constructor
- does not have to be explicitly declared in the constructor

Below is a minimal example of my issue:

```python
from sqlalchemy import Integer
from sqlalchemy.orm import Mapped, DeclarativeBase, MappedAsDataclass, mapped_column
from sqlalchemy_utc import UtcDateTime


class Base(MappedAsDataclass, DeclarativeBase):
    pass


class SoftDeleteMixin(MappedAsDataclass):
    deleted_at: Mapped[UtcDateTime | None] = mapped_column(
        UtcDateTime(),
        nullable=True,
        default=None,
    )


class Foo(SoftDeleteMixin, Base):
    __tablename__ = "Foo"
    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True, init=False)

    user_id: Mapped[int] = mapped_column(Integer)


inst = Foo(user_id=1)
print(inst)
```
This will immediately fail due to the constructor arguments.

```
TypeError: non-default argument 'user_id' follows default argument
```

I've tried to understand the relevant documentation [here](https://docs.sqlalchemy.org/en/20/orm/dataclasses.html) but I can't wrap my head around how to achieve this without dropping one my requirements. Adding `init=False` to the `deleted_at` column will fix the constructor, but prevents me from declaring the field in the constructor.  Removing the default value from `deleted_at` column would also fix the constructor, but requires explicitly declaring a value for the field on initialization. Declaring the `deleted_at` column in the main class obviously works, I'm just confused how to make it work as a mixin.

**Comment by zzzeek** [comment, marked as answer]
yeah these are hard.   We can (as I always do here, to make sure it's not us) reduce this to a plain dataclass:

```py
import dataclasses

@dataclasses.dataclass
class SoftDeleteMixin:
    deleted_at: int | None = dataclasses.field(default=None)


@dataclasses.dataclass
class Foo(SoftDeleteMixin):

    id: int = dataclasses.field(init=False)

    user_id: int = dataclasses.field()


inst = Foo(user_id=1)
print(inst)
```

same error

this is all because the way dataclasses work is that it's trying to take the list of class variables and generate an `__init__` method out of them using positional/kw rules.

When there are out-of-band field ordering problems like this, usually they can be resolved by making the oob fields kw_only, since here we of course would only use `deleted_at` as a kw argument, that resolves:

```py
import dataclasses

@dataclasses.dataclass
class SoftDeleteMixin:
    deleted_at: int | None= dataclasses.field(default=None, kw_only=True)


@dataclasses.dataclass
class Foo(SoftDeleteMixin):

    id: int = dataclasses.field(init=False)

    user_id: int = dataclasses.field()

```

and resolves for us also:

```py
from sqlalchemy import Integer, DateTime
from sqlalchemy.orm import Mapped, DeclarativeBase, MappedAsDataclass, mapped_column


class Base(MappedAsDataclass, DeclarativeBase):
    pass


class SoftDeleteMixin(MappedAsDataclass):
    deleted_at: Mapped[DateTime | None] = mapped_column(
        DateTime(),
        nullable=True,
        default=None,
        kw_only=True
    )


class Foo(SoftDeleteMixin, Base):
    __tablename__ = "Foo"
    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True, init=False)

    user_id: Mapped[int] = mapped_column(Integer)


inst = Foo(user_id=1)
print(inst)
```

**Reply by Jakkara** [reply]
Worked like a charm, thanks!
